### PR TITLE
add two API: currentProgram() and setCurrentInputMethod()

### PR DESCRIPTION
**Pull request by hosxy**
增加两个 API ，https://github.com/fcitx/fcitx5-lua/issues/7, https://github.com/fcitx/fcitx5-lua/issues/8

### CODE REVIEW

File: src/addonloader/luaaddonstate.h [hunk 1]
```diff
@@ -123,6 +123,14 @@ class LuaAddonState {
     // @function currentInputMethod
     // @treturn string the unique string of current input method.
     DEFINE_LUA_FUNCTION(currentInputMethod);
+    //
```
Comment (Aloxaf) [inline src/addonloader/luaaddonstate.h:126]: Please add document for these two function

Comment (hosxy) [reply src/addonloader/luaaddonstate.h]: Thanks, I will add it

File: src/addonloader/luaaddonstate.cpp [hunk 2]
```diff
@@ -74,6 +74,8 @@ LuaAddonState::LuaAddonState(Library &luaLibrary, const std::string &name,
             {"watchEvent", &LuaAddonState::watchEvent},
             {"unwatchEvent", &LuaAddonState::unwatchEvent},
             {"currentInputMethod", &LuaAddonState::currentInputMethod},
+            {"setCurrentInputMethod", &LuaAddonState::setCurrentInputMethod},
+            {"getCurrentProgramName", &LuaAddonState::getCurrentProgramName},
```
Comment (wengxt) [inline src/addonloader/luaaddonstate.cpp:78]: use program as name, or currentProgram

File: src/addonloader/luaaddonstate.cpp [hunk 3]
```diff
@@ -232,6 +234,22 @@ std::tuple<std::string> LuaAddonState::currentInputMethodImpl() {
...
+    if (ic) {
+        return {ic->program()};
+    }
+    return {""};
+}
+
+std::tuple<> LuaAddonState::setCurrentInputMethodImpl(const char *str) {
+    auto ic = inputContext_.get();
+    if (ic) {
+        instance_->setCurrentInputMethod(ic, str, true);
```
Comment (wengxt) [inline src/addonloader/luaaddonstate.cpp:248]: suggest to have the local parameter in the interface

File: src/addonloader/luaaddonstate.h [hunk 4]
```diff
@@ -123,6 +123,14 @@ class LuaAddonState {
     // @function currentInputMethod
     // @treturn string the unique string of current input method.
     DEFINE_LUA_FUNCTION(currentInputMethod);
+    /// Change the current input method
+    // @function setCurrentInputMethod
+    // @string str the input method name.
+    DEFINE_LUA_FUNCTION(setCurrentInputMethod);
+    /// Return the current program name
+    // @function getCurrentProgramName
+    // @treturn string the unique string of current program name.
```
Comment (wengxt) [inline src/addonloader/luaaddonstate.h:132]: program name is nothing "unique".

File: src/addonloader/luaaddonstate.h [hunk 5]
```diff
@@ -123,6 +123,14 @@ class LuaAddonState {
     // @function currentInputMethod
     // @treturn string the unique string of current input method.
     DEFINE_LUA_FUNCTION(currentInputMethod);
+    /// Change the current input method
+    // @function setCurrentInputMethod
+    // @string str the input method name.
```
Comment (wengxt) [inline src/addonloader/luaaddonstate.h:128]: this should be the "unique" name of input method, means that it's a string like "pinyin", not to be confused with the i18n'ed name

**Review by wengxt — CHANGES_REQUESTED** [review]

File: src/addonloader/luaaddonstate.cpp (hunk 3 above)
Comment (wengxt) [reply src/addonloader/luaaddonstate.cpp:248]: If you want it to accept an optional parameter, you can wrap it in base.lua

e.g.
```
local oldSetCurrentInputMethod = fcitx.setCurrentInputMethod

local function setCurrentInputMethod(name, local)
     ........
end

fcitx.setCurrentInputMethod = setCurrentInputMethod
```

local will be nil in this case, then you can do additional check, like fallback to true/false

Comment (hosxy) [reply src/addonloader/luaaddonstate.cpp:248]: 加一个 bool 类型的参数，会编译错误，貌似是因为没有定义将C++的bool类型转换为lua bool类型的规则

Comment (wengxt) [reply src/addonloader/luaaddonstate.cpp:248]: 那我建议你一起加一个，在 luahelper.h 里面，照着 int 的加一个就行了

参数需要 check，返回值需要 ret checkinterger / pushinteger 换成 checkboolean 和 pushboolean

luafunc 里面也要加上对应的函数

**Review by wengxt — CHANGES_REQUESTED** [review]

File: src/addonloader/luaaddonstate.cpp (hunk 3 above)
Comment (hosxy) [reply src/addonloader/luaaddonstate.cpp:248]: 貌似没有luaL_checkboolean()这个函数
感觉可以使用lua_toboolean()这个函数

**Comment by wengxt** [comment]
Thanks!
Please remember to run clang-format next time.

**Pull request merged**